packages/live-preview-sdk/src/inspectorMode/utils.ts

```diff
@@ -67,24 +137,25 @@ export function getAllTaggedElements(root = window.document, ignoreManual?: bool
   // FILTER_SKIP: Skip the current node
   // FILTER_REJECT: Skip the current node and all its children
   const walker = document.createTreeWalker(root, NodeFilter.SHOW_ALL, (node) => {
-    // If the node is a text node, decode its content.
-
-    if (node.nodeType === Node.TEXT_NODE) {
-      const text = node.textContent ?? '';
-
-      // Decode the text content to check for Inspector Mode CSM
-      const decoded = decode(text);
+    // Ignore tags that are not visible on the page
+    if (
+      node.nodeType === Node.ELEMENT_NODE &&
+      IGNORE_TAGS.includes((node as HTMLElement).tagName)
+    ) {
+      return NodeFilter.FILTER_REJECT;
+    }
 
-      // Skip the node if it does not have Inspector Mode CSM
-      if (decoded?.origin !== 'contentful.com') {
+    // Detect auto-tagged nodes
+    const sourceMap = getSourceMap(node);
+    if (sourceMap?.origin === 'contentful.com') {
+      // Ignore if the parent is already tagged with the same information
+      if (
+        (isTaggedElement(node.parentElement) || hasTaggedParent(node.parentElement, sourceMap)) &&
```

Comment:
Good input, I have an idea now to improve it without abandon the tree walker

Reply:
Sounds good, care to share 😃 ?

Reply:
> We should know the tagged parents already

Do we? I think that we require a separate scan because currently we get both manual & auto tagged elements with a single scan.

> I would try to persist them and then check with contains instead, which should be faster than going up the tree again 🤔

Unsure, if I got it right, you would still iterate over a list of known elements whenever you encounter an auto-tagged node? If yes, then wouldn't have eliminated the nested loop entirely.

Reply:
Yeah would not eliminate it, but should be faster than traversing the DOM up again 🤔

Reply:
Then I didn't really get your idea. From what I understood you would traverse the DOM again 😅

Reply:
will try to make a prototype today 👍 and a test case with the performance metric